pelican/plugins/pelican_redirect/__init__.py

```diff
@@ -35,28 +43,39 @@ def __init__(self, *args, **kwargs):
 
     def redirect_articles(self, article_generator):
         for article in article_generator.articles:
-            if article.metadata.get("original_url"):
-                self.redirects.append(article)
+            for url in self.historical_urls(article_generator, article, kind="ARTICLE"):
+                self.redirects.append(Redirect(to=article, from_url=url))
 
     def redirect_pages(self, page_generator):
         for page in page_generator.pages:
-            if page.metadata.get("original_url"):
-                self.redirects.append(page)
+            for url in self.historical_urls(page_generator, page, kind="PAGE"):
+                self.redirects.append(Redirect(to=page, from_url=url))
+
+    def historical_urls(self, generator, page, kind):
+        for formats in generator.settings.get("PAST_ARTICLE_URL_FORMATS", {}).values():
+            if f"{kind}_URL" in formats:
+                yield self.format_url_from_page(page, formats[f"{kind}_URL"])
+
+        if page.metadata.get("original_url"):
+            yield page.metadata.get("original_url")
```

Comment:
I chose to add them because redirects can come from multiple places, and I wanted to include the `original_url` in there. There's no real upper bound on the number of "original URLs", so why not just extend it?

In my version, a whole site map _and_ an original URL on a page-or-article-by-article basis both work.

Reply:
> There's no real upper bound on the number of "original URLs", so why not just extend it?

Fair enough! I think we just need to make sure that there are no duplicates in the final list of redirects, I'm not sure how Pelican handles overwriting files like that.

Reply:
Looking at Pelican's code, it appears that "last writer wins" is the way it goes. That seems to apply across all of the generator, so I'm not sure how we'd fix that here. If it's a blocker, it might be possible to figure it out, but I'm actually not sure there's enough information available in the generator plugins to make that decision.

Reply:
OK, this seems fine then... there shouldn't still be files at the old locations, so it's probably OK. I was thinking just to check the list of `self.redirects` for any duplicates. Since they're using dataclasses, running it through a `set` should eliminate the duplicates.